File: QA_Interview_Questions/17-group-anagrams.ts
/**
 * Question 17: Group Anagrams
 * Problem: Group strings that are anagrams together
 * 
 * Example:
 * Input: ["eat", "tea", "ate", "bat", "tab", "cat"]
 * Output: [["eat", "tea", "ate"], ["bat", "tab"], ["cat"]]
 */

function groupAnagrams(words: string[]): string[][] {
  const groups: Map<string, string[]> = new Map();
  
  for (const word of words) {
    // Sort characters to create a key
    const key = word.split('').sort().join('');
    
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    
    groups.get(key)!.push(word);
  }
  
  return [...groups.values()];
}

// Alternative: Using character count as key
function groupAnagramsCharCount(words: string[]): string[][] {
  const groups: Map<string, string[]> = new Map();
  
  for (const word of words) {
    // Create a key based on character counts
    const charCount = new Array(26).fill(0);
    
    for (const char of word) {
      charCount[char.charCodeAt(0) - 'a'.charCodeAt(0)]++;
    }
    
    const key = charCount.join(',');
    
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    
    groups.get(key)!.push(word);
  }
  
  return [...groups.values()];
}

// Test cases
console.log(groupAnagrams(["eat", "tea", "ate", "bat", "tab", "cat"]));
// Output: [["eat", "tea", "ate"], ["bat", "tab"], ["cat"]]

console.log(groupAnagrams([""]));
// Output: [[""]]

console.log(groupAnagrams(["a"]));
// Output: [["a"]]

console.log(groupAnagramsCharCount(["eat", "tea", "ate", "bat", "tab", "cat"]));
// Output: [["eat", "tea", "ate"], ["bat", "tab"], ["cat"]]

File: Java_Interview_Questions/17-GroupAnagrams.java
/**
 * Question 17: Group Anagrams
 * Problem: Group strings that are anagrams together
 * 
 * Example:
 * Input: ["eat", "tea", "ate", "bat", "tab", "cat"]
 * Output: [["eat", "tea", "ate"], ["bat", "tab"], ["cat"]]
 */

import java.util.*;

public class GroupAnagrams {
    
    // Using sorted characters as key
    public static List<List<String>> groupAnagrams(String[] words) {
        Map<String, List<String>> groups = new HashMap<>();
        
        for (String word : words) {
            char[] chars = word.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);
            
            groups.putIfAbsent(key, new ArrayList<>());
            groups.get(key).add(word);
        }
        
        return new ArrayList<>(groups.values());
    }
    
    // Using character count as key
    public static List<List<String>> groupAnagramsCharCount(String[] words) {
        Map<String, List<String>> groups = new HashMap<>();
        
        for (String word : words) {
            int[] charCount = new int[26];
            for (char c : word.toCharArray()) {
                charCount[c - 'a']++;
            }
            
            StringBuilder key = new StringBuilder();
            for (int count : charCount) {
                key.append(count).append("#");
            }
            
            String keyStr = key.toString();
            groups.putIfAbsent(keyStr, new ArrayList<>());
            groups.get(keyStr).add(word);
        }
        
        return new ArrayList<>(groups.values());
    }
    
    // Sort groups by size
    public static List<List<String>> groupAnagramsSorted(String[] words) {
        List<List<String>> result = groupAnagrams(words);
        result.sort((a, b) -> b.size() - a.size());
        return result;
    }
    
    public static void main(String[] args) {
        String[] words1 = {"eat", "tea", "ate", "bat", "tab", "cat"};
        System.out.println(groupAnagrams(words1));
        // Output: [[eat, tea, ate], [bat, tab], [cat]]
        
        String[] words2 = {""};
        System.out.println(groupAnagrams(words2));
        // Output: [[]]
        
        String[] words3 = {"a"};
        System.out.println(groupAnagrams(words3));
        // Output: [[a]]
        
        System.out.println("\nUsing Character Count:");
        System.out.println(groupAnagramsCharCount(words1));
        
        System.out.println("\nSorted by Size:");
        System.out.println(groupAnagramsSorted(words1));
    }
}
